Neon impl of ChaCha20 (better size & perf)

daverodgman opened the pull request:
## Description

Neon implementation of ChaCha20 which, depending on how it's configured, is ~500b smaller than the existing impl, or up to 4.6x faster. In its default setting it has slightly better code size than the existing impl, and is 2-2.7x faster.

Tested on all combinations of Armv7 arm/thumb2; Armv8 arm/thumb2/aarch64 on clang 14, gcc 11, and some more recent versions, plus Armv8 aarch64_be on gcc 7.5.

## PR checklist

Please remove the segment/s on either side of the | symbol as appropriate, and add any relevant link/s to the end of the line.
If the provided content is part of the present PR remove the # symbol.

- [ ] **changelog** provided
- [ ] **development PR** provided
- [ ] **framework PR** not required
- [ ] **3.6 PR** not required because: non-functional change
- [ ] **2.28 PR** not required because: non-functional change
- [ ] **tests**  to follow via separate PR in mbedtls repo - see https://github.com/Mbed-TLS/mbedtls/pull/9701/commits/a8e9a884bd5582f9d5a9adf22472380b45864d2b

| | multiblock setting | size delta c.f. development (aarch64) | perf delta c.f. development (aarch64) |
| --- | --- | --- | --- |
| clang | 0 | -84 | 0% |
| | 1 | -548 | 14% |
|  | 2 | -344 | 64% |
|  | 3 | -144 | 98% |
|  | 4 | 76 | 172% |
|  | 5 | 352 | 161% |
|  | 6 | 576 | 224% |
|  |  |  |  |
| gcc | 0 | -71 | 0% |
| | 1 | -519 | 57% |
|  | 2 | -275 | 129% |
|  | 3 | 1 | 177% |
|  | 4 | 309 | 280% |
|  | 5 | 665 | 273% |
|  | 6 | 881 | 355% |

daverodgman committed:
Minor size improvements

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

daverodgman committed:
add Changelog entry

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

daverodgman committed:
Simple Neon implementation

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

daverodgman committed:
Move xor into Neon implementation

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

daverodgman committed:
Refactor to prepare for multiblock

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

daverodgman committed:
Introduce chacha20_neon_inc_counter

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

daverodgman committed:
Tidy up Neon state into a struct

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

daverodgman committed:
Add multiblock support

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

daverodgman committed:
Code size improvement

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

daverodgman committed:
Add MBEDTLS_ARCH_IS_THUMB

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

daverodgman committed:
Adjust defaults to prevent code-size regression

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

ronald-cron-arm commented:
Thanks for moving your PR to the new repo.

davidhorstmann-arm reviewed (changes requested):
This looks good to me generally. Could you add a testcase in `tests/suites/test_suite_chacha20.data` that depends on `MBEDTLS_HAVE_NEON_INTRINSICS`? It can be a duplicate of one of the existing cases.

That way the CI should complain if we don't test this implementation at all, since any testcase that is never executed will generate an error in `analyze_outcomes.py`.

davidhorstmann-arm commented:
It also looks like this could do with a rebase on top of latest for the CI's benefit.

daverodgman committed:
Add chacha test with a dependency on MBEDTLS_HAVE_NEON_INTRINSICS to ensure testing on Neon

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

daverodgman commented:
Rebased (I just ran `git rebase` which was clean, haven't deeply inspected the result). Added requested test.

If you can advise on how/where to pull in [‎tests/scripts/components-configuration-crypto.sh](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/159#diff-1ed4ee9db4199382a2d6256c12dd611e60ceb3ba4d91b0c1c6437a0680238186) I'll add that too.

daverodgman committed:
Add testing for variations of Neon multi-block Chacha20

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

daverodgman committed:
fix changelog syntax

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

davidhorstmann-arm reviewed (commented):
Apologies for the delay, I think I have worked out why the CMake tests are not working (although I don't currently have an appropriate machine set up to test).

Would you be able to try out this change? I should then be able to approve.

daverodgman committed:
Fix use of cmake in testing different variations

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

daverodgman committed:
tiny improvement to counter increment

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

davidhorstmann-arm reviewed (dismissed):
LGTM, thanks!

daverodgman committed:
fix bug causing Neon tests to run on non aarch64 platforms

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

davidhorstmann-arm reviewed (dismissed):
LGTM, thanks!

daverodgman committed:
disable test on old versions of clang

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

davidhorstmann-arm reviewed (dismissed):
LGTM, thanks!

CI is on hold for now for release testing but I'll kick it once we are in the clear there.

daverodgman committed:
avoid compiler warning about missing null terminator

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

daverodgman committed:
make chacha20_neon_singlepass and chacha20_neon_finish_block more efficient

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

daverodgman committed:
code style

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

davidhorstmann-arm reviewed (dismissed):
LGTM, thanks!

daverodgman committed:
Merge branch 'development' into chacha-neon-split2

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

davidhorstmann-arm reviewed (dismissed):
LGTM, thanks!

yanesca commented on drivers/builtin/src/chacha20.c:
```diff
@@ -158,20 +157,22 @@ int mbedtls_chacha20_setkey(mbedtls_chacha20_context *ctx,
-    ctx->state[9]  = MBEDTLS_GET_UINT32_LE(key, 20);
-    ctx->state[10] = MBEDTLS_GET_UINT32_LE(key, 24);
-    ctx->state[11] = MBEDTLS_GET_UINT32_LE(key, 28);
+    if (MBEDTLS_IS_BIG_ENDIAN) {
```
Is there a reason for doing this check in runtime?

daverodgman replied:
Because `MBEDTLS_IS_BIG_ENDIAN` is not guaranteed to be compile-time const. On just about every reasonable compiler, it is constant, and the branch is optimised away.

yanesca commented on drivers/builtin/src/chacha20.c:
```diff
@@ -158,20 +157,22 @@ int mbedtls_chacha20_setkey(mbedtls_chacha20_context *ctx,
-    ctx->state[9]  = MBEDTLS_GET_UINT32_LE(key, 20);
-    ctx->state[10] = MBEDTLS_GET_UINT32_LE(key, 24);
-    ctx->state[11] = MBEDTLS_GET_UINT32_LE(key, 28);
+    if (MBEDTLS_IS_BIG_ENDIAN) {
```
This change is likely to make things slower if compiled with an older GCC on a platform without efficient unaligned access or if compiled with IAR on some platforms. We could prevent this if we checked for `UINT_UNALIGNED` and `UINT_UNALIGNED_STRUCT` as well.

daverodgman replied:
I would expect `memcpy` (the bit I've added) to be faster than repeated use of `MBEDTLS_GET_UINT32_LE`.

In any case, note that this is part of `mbedtls_chacha20_setkey` which is not perf sensitive. It is a decent size saving though.

yanesca commented on drivers/builtin/src/chacha20.c:
```diff
@@ -184,9 +185,13 @@ int mbedtls_chacha20_starts(mbedtls_chacha20_context *ctx,
-    ctx->state[13] = MBEDTLS_GET_UINT32_LE(nonce, 0);
-    ctx->state[14] = MBEDTLS_GET_UINT32_LE(nonce, 4);
-    ctx->state[15] = MBEDTLS_GET_UINT32_LE(nonce, 8);
+    if (MBEDTLS_IS_BIG_ENDIAN) {
```
Same as above.

yanesca commented on drivers/builtin/src/chacha20.c:
```diff
@@ -204,7 +209,7 @@ int mbedtls_chacha20_update(mbedtls_chacha20_context *ctx,
 
     /* Use leftover keystream bytes, if available */
-    while (size > 0U && ctx->keystream_bytes_used < CHACHA20_BLOCK_SIZE_BYTES) {
+    while (ctx->keystream_bytes_used < CHACHA20_BLOCK_SIZE_BYTES && size) {
```
nit (for readability):
```suggestion
    while (ctx->keystream_bytes_used < CHACHA20_BLOCK_SIZE_BYTES && size > 0) {
```

yanesca commented on drivers/builtin/src/chacha20.c:
```diff
@@ -158,20 +157,22 @@ int mbedtls_chacha20_setkey(mbedtls_chacha20_context *ctx,
-    ctx->state[2] = 0x79622d32;
-    ctx->state[3] = 0x6b206574;
+    static const char EXPAND_32_BYTE_K[16] = "expand 32-byte k";
+    memcpy(ctx->state, EXPAND_32_BYTE_K, 16);
```
I would think that on big endian machines this changes the values of the constants. What am I missing?

daverodgman replied:
Hmm. In `https://cr.yp.to/streamciphers/timings/estreambench/submissions/salsa20/chacha8/regs/chacha.c` this is written using:

```
static const char sigma[16] = "expand 32-byte k";
constants = sigma;
x->input[0] = U8TO32_LITTLE(constants + 0);
x->input[1] = U8TO32_LITTLE(constants + 4);
```

ie treat the string as a bunch of little-endian 32-bit ints. So it will write "e" into the lowest address, then "x", and so on, regardless of endianness. So unless I've missed something, this is equivalent behaviour to `memcpy`.

It does look like previously, the constant would have been written differently on big-endian, which looks like a bug.

daverodgman replied:
Sorry, no, you were right. I've fixed, and tested on big-endian aarch64. On little-endian (ie the common case) the `memcpy` path saves 20 bytes so worth retaining this.

yanesca commented on drivers/builtin/src/chacha20.c:
```diff
@@ -22,13 +22,250 @@
+    // Table look-up which results in an 8-bit rotate-left within each 32-bit element
+    const uint8_t    tbl_rotl8[16] = { 3, 0, 1, 2, 7, 4, 5, 6, 11, 8, 9, 10, 15, 12, 13, 14 };
+    const uint8x16_t vrotl8_tbl = vld1q_u8(tbl_rotl8);
+    result = vreinterpretq_u32_u8(vqtbl1q_u8(vreinterpretq_u8_u32(v), vrotl8_tbl));
```
I think this raises two questions:
1. do the performance improvements persist when the DIT bit is set (I mean is this implementation faster with DIT on than the else branch without DIT set)?
2. we haven't thought about how the library would relate to the DIT feature. Do we recommend users to set it? Are we going to add it to platform abstraction and control it ourselves? Something else? What security requirements we have for DIT code and non-DIT code? What security guarantees can we achieve and how do we communicate them?

 And I am more concerned about the second point, I don't feel we are ready to integrate code for which the security relies on DIT being set.

yanesca replied:
Never mind, I misread the code, it should be safe.

yanesca commented on drivers/builtin/src/chacha20.c:
```diff
@@ -22,13 +22,121 @@
+static inline uint32x4_t chacha20_neon_vrotlq_8_u32(uint32x4_t v)
+{
+    uint32x4_t result;
+#if defined(MBEDTLS_ARCH_IS_ARM64)
```
I am a bit concerned on the [DIT bit](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/159/commits/bd5c34ccf89b6c2033d9b31d5d1b52b51013a35d#r2303621644) requirement for this to be constant time. Do you think we could go for the `else` branch for now and add this part back when we managed to sort out our thoughts around DIT?

daverodgman replied:
Note that the look-up table that describes how to re-order the data is a compile-time const - only the register it is shuffling is unknown. So I don't think this could result in data-dependent timing.

yanesca replied:
Oh, I think I misread the code, so `tbl_rotl8` actually contains the indices and `v` is the table from which we are performing the lookup. In that case we should indeed be safe.

yanesca commented on tests/scripts/components-configuration.sh:
```diff
@@ -146,3 +146,29 @@ component_tf_psa_crypto_test_memory_buffer_allocator () {
+component_test_chacha20_neon_variations () {
+    msg "ChaCha20 Neon scalar and multiblock variations"
+
+    scripts/config.py set PSA_WANT_KEY_TYPE_AES
```
I am sorry, am not very familiar with the new config, why do we need `PSA_WANT_KEY_TYPE_AES` and `PSA_WANT_ALG_SHA_256` for ChaCha?

daverodgman replied:
This was needed previously, but seems not necessary now.

yanesca commented on tests/scripts/components-configuration.sh:
```diff
@@ -146,3 +146,29 @@ component_tf_psa_crypto_test_memory_buffer_allocator () {
     make test
 }
+
+support_test_chacha20_variations () {
```
~~I am not sure how this function is used by our test scripts, could you please help me understand where this is called from?~~

Never mind, fixed in later commit.

yanesca reviewed (commented):
Hi Dave, thank you for the PR, I had a look and I have a few questions.

daverodgman committed:
simplify test config

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

daverodgman committed:
make comparison to zero explicit

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

daverodgman committed:
fix big-endian issue

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

yanesca commented on drivers/builtin/src/chacha20.c:
```diff
@@ -22,13 +22,253 @@
+#if defined(MBEDTLS_ARCH_IS_ARM64)
+    // This implementation is slightly faster, but only supported on 64-bit Arm
+    // Table look-up which results in an 8-bit rotate-left within each 32-bit element
+    const uint8_t    tbl_rotl8[16] = { 3, 0, 1, 2, 7, 4, 5, 6, 11, 8, 9, 10, 15, 12, 13, 14 };
```
Can we please rename this variable to `idx_rotl8` or something similar, the naming really threw me off when reading the code.

yanesca commented on drivers/builtin/src/chacha20.c:
```diff
@@ -213,6 +468,37 @@ int mbedtls_chacha20_update(mbedtls_chacha20_context *ctx,
         size--;
     }
 
+#if MBEDTLS_CHACHA20_NEON_MULTIBLOCK != 0
```
Do you think we could move the neon code into its own source file? It is easier to read/maintain like that and that is how we have been doing it with AES as well.

(The two implementations seem to be nicely separated, I would hope that this is not a very big change.)

yanesca reviewed (changes requested):
Hi Dave, thank you very much for reworking the PR, I had a look at it and I have two small last asks.

daverodgman committed:
improve variable name

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

daverodgman committed:
refactor NEON impl into separate file

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

davidhorstmann-arm reviewed (dismissed):
LGTM, thanks!

Happy with the small improvements, reviewed the refactor with `--color-moved`.

daverodgman committed:
pacify check_names

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

davidhorstmann-arm reviewed (approved):
LGTM, thanks!

yanesca reviewed (approved):
LGTM, thanks

gilles-peskine-arm commented on drivers/builtin/src/chacha20_neon.c:
```diff
@@ -0,0 +1,263 @@
+static inline uint32x4_t chacha20_neon_inc_counter(uint32x4_t v)
+{
+    if (MBEDTLS_IS_BIG_ENDIAN) {
+        uint32x4_t counter_increment = { 0, 0, 0, 1 };
```
[This initializer is not portable](https://github.com/Mbed-TLS/TF-PSA-Crypto/issues/551). It works with GCC, Clang and IAR but not MSVC. @daverodgman would you mind reviewing my proposed fix? https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/552
